### ISSUE TITLE
The inconsistency of CTC API. Please optimize it.

### ISSUE THREAD

**Issue by BenjaminChoou** [tensorflow/tensorflow#42993]
**System information**
- TensorFlow version (you are using): 2.3.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state**:

The API of CTC_loss and CTC_decoder is not consistent.

1. `tf.nn.ctc_loss(labels, logits, label_length, logit_length, logits_time_major=True, unique=None, blank_index=None, name=None)`
Default blank label is 0 rather num_classes - 1, unless overridden by blank_index.

2. `tf.nn.ctc_beam_search_decoder(inputs, sequence_length, beam_width=100, top_paths=1)`
and
`tf.nn.ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True)`
This decoder treats blank label as num_classes -1 by default. And there is no option for setting blank label index.

Therefore, it could be a bit inconvenient to train a model with CTC loss and decode with the beam search decoder.
Hope someone could optimize this API.

**Will this change the current api? How?**
Original one:
`tf.nn.ctc_beam_search_decoder(inputs, sequence_length, beam_width=100, top_paths=1)`
`tf.nn.ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True)`

New one: 
`tf.nn.ctc_beam_search_decoder(inputs, sequence_length, beam_width=100, top_paths=1, blank_index=None)`
`tf.nn.ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True, blank_index=None)`

**Who will benefit with this feature?**
Researchers who are dealing with sequence recognition problems such as scene text recognition, OCR and speech recognition.

**Comment by ebrevdo** [comment]
Reassigning to Martin Wicke to decide on priority for adding this feature.

**Comment by martinwicke** [comment]
Adding the argument to the decoders is easy, I guess. Changing their default behavior, sadly, is not. So we can add the argument and have it default to num_classes-1. That's a bummer, I agree this is very ugly. 

@tensorflow/api-owners for discussion.

**Comment by yjiangling** [comment]
I find the same proble. The CTC loss in training period for TensorFlow2.x default blank label is 0 rather num_classes - 1, but for tf.nn.ctc_beam_search_decoder it still use the num_classes -1 as blank by default and there is no option to set it.

**Comment by yjiangling** [comment]
By the way, the decode function tf.nn.ctc_greedy_decoder and tf.nn.ctc_beam_search_decoder may need to add argument logits_time_major=True to keep consistent with tf.nn.ctc_loss function.

**Comment by fchollet** [comment]
@BenjaminChoou: thank you for your proposal. Could you start a RFC doc, containing end-to-end workflow examples, so that we could get feedback from potential users of this API (which is quite specialized)? Please submit the RFC to the `tensorflow/community` repo.

(for tf-api-owners)

**Comment by BenjaminChoou** [comment]
> @BenjaminChoou: thank you for your proposal. Could you start a RFC doc, containing end-to-end workflow examples, so that we could get feedback from potential users of this API (which is quite specialized)? Please submit the RFC to the `tensorflow/community` repo.
> 
> (for tf-api-owners)

This RFC will be open for comment until October 17, 2020.
| Status        |  Proposed                                            |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [295](https://github.com/tensorflow/community/pull/295) |
| **Author(s)** | Mingjie Zhou (benjamin_chou@outlook.com), |
| **Sponsor**   | François Chollet (francois.chollet@gmail.com)                 |
| **Updated**   | 2020-10-02                                           |

# Objective
Change the APIs of connectionist temporal classification (CTC) decoders (beam search decoder and greedy decoder) to ensure the consistency with the usage of CTC loss. Preserve the same arguments of API as far as possible.

**Comment by ebrevdo** [comment]
Thank you Benjamin!

**Comment by ematejska** [comment]
Thanks for filing the RFC.  Let's continue this discussion there.
